### DISCUSSION TITLE
Gleichzeitige Nutzung von CMT2300a und Huawei 4850G2

### DISCUSSION THREAD

**Discussion by darksideofthemoon1712** [hoylabs/OpenDTU-OnBattery#1132] [Q&A]
Ich habe gelesen, das sich in der Zukunft die Hardware auf ESP32 S3 ändern soll. Ist es in dem Zug auch möglich, das Cmt2300a und Huawei Charger 4850G2 über Canbus gleichzeitig verwendbar ist? Ich habe mir schon mal einen ESP32 S3 mit 16mb Flash bestellt.
Wenn das schon jemand am laufen hat wäre schön, wenn es dazu Anschlussschema und Pinmapping gibt. Da kenne ich mich leider zu wenig mit aus.

**Comment by Manos1966** [comment]
Hallo @darksideofthemoon1712 

Laut Wiki:
> The module is connected via SPI and currently requires a separate SPI bus. If you want to use the Huawei AC charger make sure to get an ESP which supports 2 SPI busses. Currently the SPI bus host is hardcoded to number 2. This may change in future. Please note: Using the Huawei AC charger in combination with the CMT2300A radio board is not supported at the moment.

Damit CMT2300a und Huawei gleichzeitig funktionieren koennen, jemand muss die Programmierung anpassen (damit CMT2300a oder Huawei ein anderes als SPI Bus 2, nutzt).

**Reply by darksideofthemoon1712** [reply]
Vielen Dank für deine Antwort, da es ja die HM Serie nicht mehr gibt und die neuen alle nur noch mit CMT2300a funktionieren, hoffe ich das es irgendwann kommt.

**Reply by Manos1966** [reply]
Ich bin genau deiner Meinung @darksideofthemoon1712 🤞 

Hast du das Problem bereits, oder handelt es sich um "langfristige Planung"?
Ich habe einen 4850G2, aber noch nicht getestet...

Mein Plan (bis eine Loesung fuer das SPI 2 Problem programmiert wird) ist, mit zwei OpenDTU-OnBattery zu arbeiten:
- erste OpenDTU-OnBattery steuert den Huawei 4850G2 (und hat Verbindung zu PowerMeter) da "the PSU power is controlled by the **Power Meter** and information provided in the web-interface". Ich brauche keine Verbindung zum HMS Wechselrichter um die Batterie zu laden.
- zweite OpenDTU-OnBattery steuert den HMS Wechselrichter (und hat Verbindung zu PowerMeter)

**Reply by darksideofthemoon1712** [reply]
Hallo,
ich bin noch in der Planung. Ich habe eine PV und ein Balkonkraftwerk, da wäre eine AC Aufladung genau das richtige.
Ich löte mir meinen ESP S3 nun mit den benötigten Komponenten auf Lochrasterplatine, dann kann ich auch nachträgliche Änderungen leicht umsetzen.  Habe jetzt eine fertige Platine und einen Powermeter über IR Lesekopf, aber aufgrund der Änderung auf größeren
Flashspeicher ja nun eine Sackgasse.
Batterie habe ich noch nicht, das Netzteil schon, weil ich ja schon ein Bisschen bauen will.
Wie begrenzt da den Einschaltstrom wenn du die Batterie auf den HMS schaltest?

**Reply by darksideofthemoon1712** [reply]
Ich habe gerade in einem Tasmota Forum gelesen das man dort mit einem TJA1051 über einen CanRX CanTX ein 4850G2 einstellt und ließt. Habe die Version schon geflasht und teste das mal, wenn ich Zeit habe. Vielleicht wäre das auch bei openDTU on Battery die Lösung.

**Reply by Manos1966** [reply]
Ich bin auch gerade "am spielen"
![Huawei Step2small](https://github.com/user-attachments/assets/b4ea5345-a4e9-4e22-acc3-c0b5fdb7ff5a)

Hast du ein Link fuer deine Version?

Ich benutze die Steuerung von KlausLi https://github.com/KlausLi/Esp-HuaweiR4850-Controller

**Reply by Manos1966** [reply]
> Wie begrenzt da den Einschaltstrom wenn du die Batterie auf den HMS schaltest?

ich habe eine Pylontech, sie hat "Soft-Start" Option

**Comment by skippermeister** [comment]
Das Problem des hard codierten SPI Busses habe ich in meinem Fork gelöst. Ich habe dazu einen SPIPortManager ähnlich dem SerialPortManager implementiert. Die SPI Resourcen werden auf Anforderung den genutzen Gerätetreibern zugewiesen. Somit wird dem CMT2300A ein andere SPI Bus als dem MCP2515 für das HUAWEI Netzteil zugewiesen.

Die Änderungen und Anpassungen sind überschaubar.
SPIPortManager.cpp und .h File erstellen.
Aktivierung des Managers in main.cpp
Einbindung der SPIPortManager Funktionen in den Modulen InverterSettings.cpp (für NRF24 und CMT2300A), huawei_can.cpp.
Und in der Hoymiles und CMT2300A lib die Übergabe der SPI Nummer an das CMT2300A Modul (hier sind ein paar mehr Änderungen zu machen).

Wie gesagt in meinem Fork ist das implementiert und funktioniert.

@schlimmchen : kannst Du Dir das mal in meinem Fork anschauen.
https://github.com/skippermeister/OpenDTU-OnBattery/blob/development/src/SPIPortManager.cpp
https://github.com/skippermeister/OpenDTU-OnBattery/blob/development/include/SPIPortManager.h
https://github.com/skippermeister/OpenDTU-OnBattery/blob/development/src/main.cpp (Zeilen 14 und 106)
https://github.com/skippermeister/OpenDTU-OnBattery/blob/development/src/InverterSettings.cpp (Zeilen 11, 60 - 62 und 70 - 72)
https://github.com/skippermeister/OpenDTU-OnBattery/blob/development/src/Huawei_can.cpp (Zeilen 14 und 39 - 43)
https://github.com/skippermeister/OpenDTU-OnBattery/blob/development/lib/Hoymiles/src/Hoymiles.cpp (Zeilen 39 und 41)
https://github.com/skippermeister/OpenDTU-OnBattery/blob/development/lib/Hoymiles/src/HoymilesRadio_CMT.cpp (Zeilen 88 und 92)
https://github.com/skippermeister/OpenDTU-OnBattery/blob/development/lib/Hoymiles/src/HoymilesRadio_CMT.h (Zeile 44)
https://github.com/skippermeister/OpenDTU-OnBattery/blob/development/lib/CMT2300a (diverse Sourcecode Dateien, suche spi_host)

**Reply by darksideofthemoon1712** [reply]
Hallo Skippermeister, das ist wirklich eine tolle Leistung. Genau so etwas suche ich.
Hast du vielleicht noch ein Pinmapping für meinen ESP32 S3 USB?

![ESP32 S3 USB](https://github.com/user-attachments/assets/66f49c52-6d41-4431-ab2d-c2d82d5db7a8)

**Reply by AndreasBoehm** [reply]
Danke @skippermeister.
Da dieses Thema oft angefragt wird würde ich mich die Tage drum kümmern deinen commit (https://github.com/skippermeister/OpenDTU-OnBattery/commit/a642c9dbdb51d827cb7e709e58272ab682f4f680) hier ins repo zu integrieren.

**Reply by schlimmchen** [reply]
Wie löst denn der SPI port manager das bekannte Problem? Wie viel SPI Busse gibt es, und welcher macht "normalerweise" was? Ich dachte, es sei davon eh nur einer von frei, und der kann dann entweder CMT2003 *oder* Huawei CAN. Verstehe ich etwas falsch?

**Reply by AndreasBoehm** [reply]
Das Problem ist das CMT2300 und Huawei CAN beide hardcoded SPI2 nutzen, bei neueren ESPs stehen aber 2 SPIs zur freien verfügung (SPI2 und SPI3). Das müssen wir dann nur auch im wiki erwähnen.

**Reply by Manos1966** [reply]
>  Das müssen wir dann nur auch im wiki erwähnen.

z.Zeit haben wir die Erklaerung
https://github.com/helgeerbe/OpenDTU-OnBattery/blob/development/docs/hardware_flash.md#cmt2300a-radio-board--see-inverter-table-above-for-supported-inverters
"The CMT2300A uses 3-Wire half duplex SPI communication. Due to this fact it currently requires a separate SPI bus. If you want to run the CMT2300A module on the same ESP32 as a NRF24L01+ module or a PCD8544 display make sure you get a ESP which supports 2 SPI busses. Currently the SPI bus host is hardcoded to number 2. **This may change in future.**"

sowie den Standard Spruch
"Using the Huawei AC charger in combination with the CMT2300A radio board (for HMT- HMS- Inverters) is not supported **at the moment**"

https://github.com/helgeerbe/OpenDTU-OnBattery/wiki/FAQ#huawei-ac-charger-rg4850-with-cmt2300a-radio-board-for-hmt--hms--series-inverters

https://github.com/helgeerbe/OpenDTU-OnBattery/wiki/Device-Profiles#other-information

https://github.com/helgeerbe/OpenDTU-OnBattery/wiki/Huawei-AC-PSU#other-observations

Wenn es soweit ist, kann ich entsprechend anpassen...

**Reply by skippermeister** [reply]
Hallo @darksideofthemoon1712

Für die original OpenDTU-onBattery könnte nach Implementierung des SPIPortManagers ein Pinmapping für ein ESP32-S3-USB mit
- CMT2300A
- HUAWEI
- Batterie SmartShunt (VE.Direct) oder Pylontech (CAN) oder Pytes (CAN) oder JKBMS
- 2x Victron MPPT
- RS485 für Powermeter (SoftwareSerial mit DERE)
- SH1106 Display (I2C)

wie folgt aussehen [ESP32_S3_USB_CMT_HUAWEI.json](https://github.com/user-attachments/files/16418321/ESP32_S3_USB_CMT_HUAWEI.json)

**Reply by AndreasBoehm** [reply]
@skippermeister konntest du den SPIPortManager bereits mit NRF24 und CMT im parallel betrieb testen? Hab ich mit einem deiner release builds und auch mit deinem code portiert durch mich nach onBattery nicht zum laufen bekommen, die schliessen sich gegenseitig aus. Huawei kann ich leider nicht testen, hab keinen und auch keinen MCP2515.

Ich vermute das es daran liegt das SP0 und SPI1 fürs system reserviert sind, wir die aber auch gar nicht ansprechen können und wir deshalb bei index 0 anfangen können. So sah es zumindest nach einem schnellen versuch mit meiner codebase aus.

Hab einen ESP32S3.

**Reply by skippermeister** [reply]
Hallo @AndreasBoehm
Ich habe leider keinen CMT, jedoch habe ich beim Initialisieren des SPI Ports für den NRF24 und dem CMT keinen Resourcen Fehler bekommen.
Eventuell hast Du noch bei der Übernahme meines Commits etwas vergessen. Die SPI Nummer beim NRF24 und beim MCP2515 ist grundsätzlich einen Zähler höher als die SPI Host Nummer. Die Initialisierungsroutine für den CMT arbeitet mit anderen Kernal Routinen als die Initialisierung des NRF24 bzw. des MCP2515.
Ich arbeite mit einem ESP32. Hast Du einen ESP32 S3 ?
Bei Indes 0 oder 1 darfst Du nicht anfangen. Ich habe deshalb im SPIPortManager Index0 und Index1 für Flash und PSRAM reserviert.
Für NRF24 und MCP2515 fängt der Index bei 2 an. Für den CMT musst Du bei 1 anfangen, also den Returnwert des SPIPortManagers um 1 subtrahieren.

Für NRF24 gilt:
SPIClass* spiClass = new SPIClass(***oSPInum**);

Für CMT gilt:
Hoymiles.initCMT(***oSPInum - 1,** pin.cmt_sdio, pin.cmt_clk, pin.cmt_cs, pin.cmt_fcs, pin.cmt_gpio2, pin.cmt_gpio3);


Zeige mir mal in Deiner Implementierung ein Log vom Start des ESP32 bis etwa 1min nach Start.

**Reply by Manos1966** [reply]
@AndreasBoehm @skippermeister 
ich werde einen HMS Ende der Woche bestellen.
Ein paar HM- habe ich sowieso, Pylontech, Victron, Huawei und ein paar ESP32, ESP32-S3 ebenfalls. 
Von Programmierung habe ich keine Ahnung (das letzte Mal, als ich Code geschrieben habe, war Anno 1988 🙊), aber ich kann helfen beim Testen falls gewünscht ist.

**Reply by AndreasBoehm** [reply]
@skippermeister leider konnte ich mit VSCode keine Logs finden in denen der SPIPortManager auftaucht.

Ich habe aber etwas interessantes in `esp32-hal-spi.h` gefunden. Die unterschiedliche Nummerierung der SPIs passt perfekt zu dem Verhalten das ich mit meinem S3 feststellen konnte und erklärt auch warum es auf deinem ESP32 wunderbar funktioniert.

```c++
#if CONFIG_IDF_TARGET_ESP32C3 || CONFIG_IDF_TARGET_ESP32S3
#define FSPI  0
#define HSPI  1
#else
#define FSPI  1 //SPI bus attached to the flash (can use the same data lines but different SS)
#define HSPI  2 //SPI bus normally mapped to pins 12 - 15, but can be matrixed to any pins
#if CONFIG_IDF_TARGET_ESP32
#define VSPI  3 //SPI bus normally attached to pins 5, 18, 19 and 23, but can be matrixed to any pins
#endif
#endif
```

Für CMT wurde, wie du schon sagtest der HOST `SPI2_HOST` definiert und hier gibt es keine unterschiede zwischen ESP32 und S3. Deshalb must du `*oSPInum - 1` nehmen und für mich hat es nur geklappt wenn ich `*oSPInum + 1` nutze.

Auszug aus `spi_types.h`
```c++
typedef enum {
//SPI1 can be used as GPSPI only on ESP32
    SPI1_HOST=0,    ///< SPI1
    SPI2_HOST=1,    ///< SPI2
#if SOC_SPI_PERIPH_NUM > 2
    SPI3_HOST=2,    ///< SPI3
#endif
} spi_host_device_t;
```

Ich denke das mapping sieht aus wie hier in der Tabelle

**ESP32-S3**
 | SPI # | SPI ID | SPI_HOST |
|---|---|---|
 | 0     | N/A     | 0        |  
| 1     | 0      | 0       |  
 | 2     | 1      | 1        |  
 | 3     | 3      | 2        |  

**ESP32**
 | SPI # | SPI ID | SPI_HOST |
|---|---|---|
 | 0     | N/A      | 0        |  
 | 1     | 1      | 0       |  
| 2     | 2      | 1        |  
 | 3     | 3      | 2        |  

- `SPI0 and SPI1 are used internally to access the ESP32’s attached flash memory. Both controllers share the same SPI bus signals.`
- `SPI2 and SPI3 are general purpose SPI controllers, sometimes referred to as HSPI and VSPI, respectively. They are open to users. SPI2 and SPI3 have independent bus signals with the same respective names.`

_Auszug von [hier](https://docs.espressif.com/projects/esp-idf/en/v4.4/esp32/api-reference/peripherals/spi_master.html)_


Mein draft pull-request ist hier zu finden: https://github.com/helgeerbe/OpenDTU-OnBattery/pull/1144
Unterstützt aktuell nur ESP32-S3

**Reply by skippermeister** [reply]
@AndreasBoehm 
der SPIPortManager funktioniert aktuell anscheinend nur mit ESP32. Die SPI Port Nummern werden im ESP32S3 anders nummeriert. Ich habe mir eben einen ESP32S3 16MB RAM / 8MB PSRAM programmiert.  (übrigens wird bei mir der PSRAM nicht erkannt, das muss etwas mit den unterschiedlichen PSRAM Line Modes zu tun habe, da werde ich auch noch forschen müssen, ist aber erst einmal zweitrangig).

Änderungen habe ich programmiert. Die ersten Tests mit dem ESP32S3 und NRF24 zeigen keinen Fehler. 
Den nächsten Test baue ich auf einem Steckboard mit dem ESP32S3, einem NRF24 und MCP2515 auf.
Ein CMT2300A Modul habe ich mir gestern noch bestellt.

Die aktuelle Änderung habe ich hochgeladen (Source Code), [fix: SPIPortManager für ESP32S3 (1.Test)](https://github.com/skippermeister/OpenDTU-OnBattery/commit/a2c91efe409a9dde2748db316dbf859d4fd33503)

Wer kann das mit einem NRF24 und CMT2300A testen.

**Reply by skippermeister** [reply]
schau mal mein Source Code Fix an. Ich habe das jetzt so geändert, dass in ESP32 und ESP32S3 im Compilerlauf durch Macros unterschieden wird. Ich bin gestern Abend auf die gleichen Ergebnisse gekommen.

**Reply by AndreasBoehm** [reply]
ESP32-S3 sieht gut aus, NRF24 und CMT laufen jetzt gleichzeitig. Logs sehen auch gut aus. Danke.

**Reply by AndreasBoehm** [reply]
In meinem MR gibts jetzt einen build zum ausprobieren: https://github.com/helgeerbe/OpenDTU-OnBattery/pull/1144#issuecomment-2261338767

**Reply by schlimmchen** [reply]
MR? Aha, noch einer der sonst gitlab nutzt statt github 😁

**Reply by AndreasBoehm** [reply]
@schlimmchen hast mich erwischt 😂 Ich meine natürlich PR.
